### ISSUE TITLE
MobileIAP "Complete product registration" action doesn't "trigger" anymore...?

### ISSUE THREAD

**Issue by DaveedA13** [Scirra/Construct-3-bugs#3523]
## Problem description

MobileIAP "Complete product registration" action doesn't "trigger" anymore...?

Or am I crazy...?

If this action is in my "on start layout", just one time in my whole project.

Either "On store registration success" OR "On store registration failed" should trigger...?

I'm pretty sure it worked up until recently.

I even retried my minimal project in the latest stable C3, could it be a problem at export time "server-side"...?

Am I still doing my IAP events the wrong way...?  

## Attach a .c3p

[Report Build 7 MobIAP in R171.3 - Z.zip](https://github.com/Scirra/Construct-3-bugs/files/3922577/Report.Build.7.MobIAP.in.R171.3.-.Z.zip)

## Steps to reproduce

1. Open the project (I tried it in R171.3 but had the problem in latest beta release too).
2. Export it for XCode
3. Archive it with XCode
4. Test it in Test Flight.

For info, the "buy" button do work in TEST FLIGHT and the "on product purchase success" triggers too, then the button is disabled by the events.

...but ideally I want to show the button only if store registration was a success, same for showing a "Restore purchase" button...

## Observed result

No debug information for any of the 2 triggers...

## Expected result

Debug information text should show "success" or "failed"...

## More details

**Affected browsers/platforms:** iOS app.

**First affected release:** Worked 2 weeks ago approximately, sorry for not being sure. I worked a few days on trying to find what went wrong...by recreating all my events in old backup of my c3 project...

## System details

<details><summary>View details</summary>

Platform information
Browser: NW.js
Browser version: 0.39.0
Browser engine: Chromium
Browser architecture: 64-bit
Context: nwjs
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1.25
Logical CPU cores: 4
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (@ac9418ba9c3bd7f6baaffa0b055dfe147e0f8364) NWjs/0.39.0 (like Chrome/75.0.3770.80)
C3 release: r178 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 8 gb
Storage usage (approx): 1.1 gb (13.3%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (NVIDIA GeForce GT 630 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 44100 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
</details>

**Comment by DaveedA13** [comment]
Sorry for the new post but...any update on this @shortercode ...?

Am I right that "Complete product registration" action should trigger a success or failed condition...?

**Comment by shortercode** [comment]
Your right that complete product registration should fire after a successful registration. Unfortunately the failure trigger doesn't really work anymore, as the plugin retries until it successfully loads. This is to allow the plugin to work when the device is initially offline.

The last update to the cordova plugin was about 2 months ago. As for the Construct plugin there was a very minor change in early November that prevents attempts at purchases an invalid product, but that shouldn't have any effect on the registration completion.

**Comment by DaveedA13** [comment]
Thanks for the answer @shortercode  about the "failure trigger", noted and understood. :)

...but...so, if it doesn't work, we can't get an error message about it...? ...to debug...

Would you still my provided minimal c3p file here later on...? I would like to know if my events seems ok.

I'm the one that pointed out the "purchase invalid product" crash report, and I'm thankful that you corrected it. :)

I'm a little lost here, I can't make my "purchase" button visible in my game since I'm checking if the store registration was completed before...and it doesn't trigger.

**This part always worked fine before**, even when I had the "crash" on "purchase invalide product". :/   ...even in TEST FLIGHT app I got "store registration" to "complete", but not now anymore...

**Comment by shortercode** [comment]
Yes I'm aware that it's not particularly useful that the failure trigger doesn't work, best thing is just to treat not loaded as a failure. It's common for the IAP page to not be the first in your application; so if you setup as soon as your app loads, then chances are it **should** have loaded by the time your user has reached the store page. If it hasn't then show a generic "unable to load store" error to them.

I haven't been able to reproduce any issues here, and as far as I can tell your events are correct. It might be that the store on your device has ended up in an odd state. But that would be very hard for us to debug. I think it's more likely a configuration issue on itunes connect/etc.

If it is related to the state of the store, we have made some updates since you have filed this. So you might have more luck with the latest beta release?

**Comment by DaveedA13** [comment]
Thanks for the answer @shortercode !

Ok for the tips about the "treat not loaded as a failure".

Big thank you for the confirmation about my events too, was getting crazy...

I didn't get back here since a few weeks since I knew that you guys were on holidays...

I tried with the latest beta release and still no change, I tested on 2 different devices.

I would like to open a new report with a new minimal file to test if you don't mind, to me it seems there is somethign going wrong with the "store registration success" because it "fires" only when no purchase was made so far... (That's what I discovered, and it's why I tought that it didn't trigger anymore)

Is "store registration success" a condition that can trigger only once per device...?

Even with a "consumable" IPA, I can buy it only once, with a minimal file created by Construct 3 dev team.

I would really appreciate if you could take a look at my next minimal project and try it too.

Thanks in advance.

**Comment by AshleyScirra** [comment]
This issue is marked 'can't reproduce' and it appears a follow-up issue was filed at #3608, so closing this one.
